<a href="https://colab.research.google.com/github/Viraj1901181/CF969-Big-Data-for-Computational-Finance/blob/master/LAB%208/Lab_8_Neural_Networks_in_TensorFlow.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#CF969 - Big Data for Computational Finance

##Lab 8: Neural Networks in TensorFlow

In the previous lab you have learned some of the basics of TensorFlow, and in particular you have worked with autodiff and gradient descent.

In this week's lab you will see how to use those tools for building and training some basic neural networks in TensorFlow

The present Notebook for this lab is based on a previous version written by Bart de Keijzer, and also on a tutorial and a big set of slides created by Dr. Michael Fairbank: he is the creator of most of the material that you see here. Some of the Python code scripts here are based on code from tensorflow.org and from https://github.com/aymericdamien/TensorFlow-Examples/, by Aymeric Damien.

When going through these notes, you should work, as usual, to work on a new Jupyter Notebook where you recreate and re-execute all the code snippets in the present notebook. Please experiment with the pieces of code, take your time with them, and look up the meaning of the various statements in the online TensorFlow documentation whenever you do not fully understand what is happening.

##A Neural Network in Only a Few Lines of Code

TensorFlow has some of the basic and common neuron activation functions built-in. A tensor can be passed to such a function, after which said function is applied to each element of the tensor. In particular, there is the function sigmoid(A) which applies the sigmoid function to each element of tensor A.

While in our theoretical treatment of neural networks, we usually represent a network's input by a column vector, in TensorFlow we will use a row vector to represent the input. We do this for ease of programming. Recall that for a sigmoid neuron, the sigmoid function is applied on a value that is a weighted sum of all inputs, plus a bias. Therefore, if we now represent the weights by a column vector w and the bias by a scalar b (i.e. rank 0 tensor), we can now write the application of a sigmoid neuron to the input vector x using the command tf.sigmoid(tf.matmul(x,w)+b) . Note however that for matmul to work, we must explicitly define x and w as rank 2 tensors, despite that they can be more naturally seen as rank 1 tensors. This is simply because matmul expects its argument to be of rank 2.

In [1]:
import tensorflow as tf
sess = tf.Session()
x=tf.constant([[1,1]], tf.float32) # our input vector
w=tf.constant([[2],[1]],tf.float32) # weight vector of a sigmoid neuron
b=tf.constant(1,tf.float32) # bias of a sigmoid neuron
y=tf.sigmoid(tf.matmul(x,w) + b)
print(sess.run(y))

[[0.98201376]]


We can generalise this easily to applying the input to multiple sigmoid neurons (that is, a layer of neurons). Let A be a matrix where the columns are the weight vectors of all the neurons. Let b be a row vector of all the biases of the neurons. The command tf.sigmoid(tf.matmul(x,A)+b) then results in a row vector containing the output of each of the neurons.

Exercise: Take the above code and modify it by adding a second neuron with weights 1 and -1, and with a bias of -0.5. The answer should be a row vector with the values 0.98201376 and 0.37754068.

In [2]:
import tensorflow as tf
sess = tf.Session()
x=tf.constant([[1,1]], tf.float32) # our input vector
w=tf.constant([[1],[-1]],tf.float32) # weight vector of a sigmoid neuron
b=tf.constant(0.5,tf.float32) # bias of a sigmoid neuron
y=tf.sigmoid(tf.matmul(x,w) + b)
print(sess.run(y))

[[0.62245935]]
